docs/cuda-custom-call/mul.cu

```diff
@@ -0,0 +1,42 @@
+#include <xla/ffi/api/c_api.h>
+#include <xla/ffi/api/api.h>
+#include <xla/ffi/api/ffi.h>
+
+// an elementwise product with a grid stride loop
+__global__ void MulKernel(const float *a, const float *b, float *c, size_t n) {
+  size_t tid = blockIdx.x * blockDim.x + threadIdx.x;
+  const size_t grid_stride = blockDim.x * gridDim.x;
+  for (size_t i = tid; i < n; i += grid_stride)
+    c[i] = a[i] * b[i];
+}
+
+// host wrapper function that launches the kernel with hardcoded grid/block size
+void LaunchMulKernel(cudaStream_t stream,
+                     const float *a, const float *b, float *c, size_t n) {
+  const int block_dim = 128;
+  const int grid_dim = 1;
+  MulKernel<<<grid_dim, block_dim, /*dynamic_shared_mem_bytes=*/0, stream>>>(
+      a, b, c, n);
+}
+
+namespace ffi = xla::ffi;
+
+// XLA FFI binding wrapper around our host launcher function
+extern "C" XLA_FFI_Error *Mul(XLA_FFI_CallFrame *call_frame) {
+  static const auto *kImpl =
+      ffi::Ffi::Bind()
+          .Ctx<ffi::PlatformStream<cudaStream_t>>()
+          .Arg<ffi::Buffer<ffi::DataType::F32>>()
+          .Arg<ffi::Buffer<ffi::DataType::F32>>()
+          .Ret<ffi::Buffer<ffi::DataType::F32>>()
+          .Attr<size_t>("n")
+          .To([](cudaStream_t stream,
+                 ffi::Buffer<ffi::DataType::F32> a,
+                 ffi::Buffer<ffi::DataType::F32> b,
+                 ffi::Result<ffi::Buffer<ffi::DataType::F32>> c,
+                 size_t n) -> ffi::Error {
+            LaunchMulKernel(stream, a.data, b.data, c->data, n);
+            return ffi::Error::Success();
+      }).release();
+  return kImpl->Call(call_frame);
+}
```

Comment:
Not crucial, but some of this boilerplate could be abstracted with the [`XLA_FFI_DEFINE_HANDLER` macro](https://github.com/openxla/xla/blob/6d17fd8fbe27b06fc4828dc0c38ecdd8f9ba8573/xla/ffi/api/api.h#L1602). This is the syntax used in the [XLA custom call docs](https://github.com/openxla/xla/blob/6d17fd8fbe27b06fc4828dc0c38ecdd8f9ba8573/docs/custom_call.md) now.

Reply:
Thanks, this looks like this should reduce the amount of boilerplate. The flip side is I'm now looking at a template error that I can't understand :)

Reply:
The error was due to the fact that the function I was exporting using this macro has return type `void`, but it needs to be an `ffi::Error` instead.

But now I'm facing a different issue: the macro creates a local lambda variable instead of a function definition. This doesn't work for my case because I'm trying to export a function with C linkage.

Reply:
One solution suggested by @ezhulenev is to add a new macro that defines a function with C linkage.

Reply:
Yeah, basically `XLA_FFI_EXPORT_HANDLER` that will hide the `extern "C" XLA_FFI_Error *Mul` boilerplate. Another options is to change default behavior of `XLA_FFI_DEFINE_HANDLER` to define C function by default and add another one for users that prefer lambda.